src/physics/ra_driver.f90

```diff
@@ -242,6 +267,37 @@ subroutine rad(domain, options, dt, halo, subset)
                 last_model_time = domain%model_time%seconds()
                 domain%tend%th_swrad = 0
                 domain%shortwave%data_2d = 0
+                ! Calculate cloud fraction
+                If (options%parameters%icloud == 3) THEN
+                    DO j = jts,jte
+                        DO i = its,ite
+                            DO k = kts,kte
+                                p_1d(k) = domain%pressure%data_3d(i,k,j) !p(i,k,j)
+                                t_1d(k) = domain%temperature%data_3d(i,k,j)
+                                qv_1d(k) = domain%water_vapor%data_3d(i,k,j)
+                                qc_1d(k) = domain%cloud_water_mass%data_3d(i,k,j)
+                                qi_1d(k) = domain%cloud_ice_mass%data_3d(i,k,j)
+                                qs_1d(k) = domain%snow_mass%data_3d(i,k,j)
+                                Dz_1d(k) = domain%dz_interface%data_3d(i,k,j)
+                                cf_1d(k) = cldfra(i,k,j)
+                            ENDDO
+                            gridkm = domain%dx*1000
+                            XLAND = domain%land_mask
+                            CALL cal_cldfra3(cf_1d, qv_1d, qc_1d, qi_1d, qs_1d, Dz_1d, &
+                 &                          p_1d, t_1d, XLAND(i,j), gridkm,        &
+                 &                          .false., 1.5, kts, kte)
+
+                            DO k = kts,kte
+                                qc(i,k,j) = qc_1d(k)
```

Comment:
what is the purpose of the 3D qc,qi,qs variables? do they get updated by cal_cldfra3?